### ISSUE TITLE
JSONpath not drilling down into an RS .spec.template.metadata.name

### ISSUE THREAD

**Issue by TerraTech** [kubernetes/kubernetes#36391]
**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5", GitCommit:"5a0a696437ad35c133c0c8493f7e9d22b0f9b81b", GitTreeState:"clean", BuildDate:"2016-10-29T01:38:40Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"4", GitVersion:"v1.4.5+coreos.0", GitCommit:"f70c2e5b2944cb5d622621a706bdec3d8a5a9c5e", GitTreeState:"clean", BuildDate:"2016-10-31T19:16:47Z", GoVersion:"go1.6.3", Compiler:"gc", Platform:"linux/amd64"} 
```

**Environment**:
CoreOS-1185.3.0 (baremetal)
hyperkube-v1.4.5_coreos.0

**What happened**:
Given the following output:
```sh
$ k get rs -ojson
* I snipped out all but .item[0] for clarity
```
```json
{
    "kind": "List",
    "apiVersion": "v1",
    "metadata": {},
    "items": [
        {
            "kind": "ReplicaSet",
            "apiVersion": "extensions/v1beta1",
            "metadata": {
                "name": "bb-m01-1751755744",
                "namespace": "default",
                "selfLink": "/apis/extensions/v1beta1/namespaces/default/replicasets/bb-m01-1751755744",
                "uid": "4bf56456-a4ea-11e6-85f3-0018f3531b4f",
                "resourceVersion": "30715352",
                "generation": 1,
                "creationTimestamp": "2016-11-07T13:01:28Z",
                "labels": {
                    "fq-app": "bb-m01",
                    "pod-template-hash": "1751755744"
                },
                "annotations": {
                    "deployment.kubernetes.io/desired-replicas": "1",
                    "deployment.kubernetes.io/max-replicas": "2",
                    "deployment.kubernetes.io/revision": "2"
                }
            },
            "spec": {
                "replicas": 1,
                "selector": {
                    "matchLabels": {
                        "fq-app": "bb-m01",
                        "pod-template-hash": "1751755744"
                    }
                },
                "template": {
                    "metadata": {
                        "name": "bb-m01",
                        "namespace": "default",
                        "creationTimestamp": null,
                        "labels": {
                            "fq-app": "bb-m01",
                            "pod-template-hash": "1751755744"
                        }
                    },
                    "spec": {
                        "containers": [
                            {
                                "name": "busybox",
                                "image": "busybox",
                                "command": [
                                    "/bin/sleep",
                                    "2592000"
                                ],
                                "resources": {},
                                "terminationMessagePath": "/dev/termination-log",
                                "imagePullPolicy": "IfNotPresent"
                            }
                        ],
                        "restartPolicy": "Always",
                        "terminationGracePeriodSeconds": 30,
                        "dnsPolicy": "ClusterFirst",
                        "nodeSelector": {
                            "fq/minion": "01"
                        },
                        "securityContext": {},
                        "hostname": "bb-m01"
                    }
                }
            },
            "status": {
                "replicas": 1,
                "fullyLabeledReplicas": 1,
                "readyReplicas": 1,
                "observedGeneration": 1
            }
        },
[[..snip..]]
} 
```

---
This all started when I wanted to cleanup some straggling RS left behind by deployment updates:
```sh
$ kubectl get rs -ojsonpath='{.items[?(@.spec.replicas==0)].metadata.labels}'
error executing jsonpath "{.items[?(@.spec.replicas==0)].kind}": incompatible types for comparison

```
Which leads me to believe that JSONpath can't test integers. :disappointed:

Then I went tinkering around with other JSONpath tests and stumbled on the following problem that can be replicated with:
```sh
#!/bin/sh
echo '$ 'kubectl get rs -ojsonpath='{.items[?(@.metadata.name=="bb-m01-1751755744")].metadata.labels}'
kubectl get rs -ojsonpath='{.items[?(@.metadata.name=="bb-m01-1751755744")].metadata.labels}';echo

echo
echo '$ 'kubectl get rs -ojsonpath='{.items[?(@.spec.selector.matchLabels.pod-template-hash=="1751755744")].metadata.labels}'
kubectl get rs -ojsonpath='{.items[?(@.spec.selector.matchLabels.pod-template-hash=="1751755744")].metadata.labels}';echo

echo
echo '$ 'kubectl get rs -ojsonpath='{.items[?(@.spec.template.metadata.name=="bb-m01")].metadata.labels}'
kubectl get rs -ojsonpath='{.items[?(@.spec.template.metadata.name=="bb-m01")].metadata.labels}';echo 
```
The output from the run:
```
$ kubectl get rs -ojsonpath={.items[?(@.metadata.name=="bb-m01-1751755744")].metadata.labels}
map[pod-template-hash:1751755744 fq-app:bb-m01]

$ kubectl get rs -ojsonpath={.items[?(@.spec.selector.matchLabels.pod-template-hash=="1751755744")].metadata.labels}
map[fq-app:bb-m01 pod-template-hash:1751755744]

$ kubectl get rs -ojsonpath={.items[?(@.spec.template.metadata.name=="bb-m01")].metadata.labels}
Error executing template: name is not found. Printing more information for debugging the template:
	template was:
		{.items[?(@.spec.template.metadata.name=="bb-m01")].metadata.labels}
	object given to jsonpath engine was:
	[[..snip..]]

error executing jsonpath "{.items[?(@.spec.template.metadata.name==\"bb-m01\")].metadata.labels}": name is not found
```
I don't understand why I was able to drill down on _@.spec.selector.matchLabels.pod-template-hash_, but I couldn't for _@.spec.template.metadata.name_ since they are at the same level in the RS structure.

At this point, I don't know if this is a bug, or if I am misunderstanding something fundamental to JSONpath.

**Comment by AdoHe** [comment]
@TerraTech the JSONPath implementation did have some problems, I would take a look asap.

**Comment by fraenkel** [comment]
This is a pretty common issue since we are converting the JSON to a map[string]interface{}, all numbers are treated as float64. The template engine unfortunately documents that it compares like types which will never work out in our case.  Since we vendor & fork the template, it is a trivial change to make this all work.

**Comment by TerraTech** [comment]
Was there a resolution to this issue?

**Comment by TerraTech** [comment]
@adohe @fraenkel Is this still on your radar?

**Comment by TerraTech** [comment]
/remove-lifecycle stale

@adohe @fraenkel I'm trying to prevent this from going full necro.

**Comment by drekle** [comment]
I am also experiencing this with creationTimestamp queries

**Comment by thockin** [comment]
Digging for other JSONPath bugs, I found this.  

These seem to work:

`kubectl get rs --all-namespaces -o jsonpath="{range .items[?(@.spec.replicas==0)]}{.metadata.namespace}/{.metadata.name}{'\n'}{end}"`

`kubectl get rs --all-namespaces -o jsonpath="{range .items[?(@.spec.replicas==0)]}{.metadata.labels}{'\n'}{end}"`

Also this:

`kubectl get rs --all-namespaces -o jsonpath="{range .items[?(@.spec.template.spec.hostname=='foobar')]}{.metadata.labels}{'\n'}{end}"`

So I think this works if you use the fully-formed syntax.  I don't know why the looser syntax works.  I suggest we close this as WAI

**Comment by liggitt** [comment]
The integer comparison was fixed in 1.6 (and remains fixed in master):

```
$ kubectl version --client
Client Version: version.Info{Major:"1", Minor:"5+", GitVersion:"v1.5.9-beta.0-dirty", GitCommit:"f35802d3a00b37a32476451266af05ce9760fec0", GitTreeState:"dirty", BuildDate:"2020-05-06T15:02:40Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"darwin/amd64"}

$ kubectl get rs -ojsonpath='{.items[?(@.spec.replicas==1)].metadata.name}'
Error executing template: incompatible types for comparison. Printing more information for debugging the template:
	template was:
		{.items[?(@.spec.replicas==1)].metadata.name}
	object given to jsonpath engine was:
		map[string]interface {}{"items":[]interface {}{map[string]interface {}{"metadata":map[string]interface {}{"ownerReferences":[]interface {}{map[string]interface {}{"blockOwnerDeletion":true, "controller":true, "kind":"Deployment", "name":"my-dep", "uid":"080eb26a-ec5a-4cbc-b15d-aa63a3d9d252", "apiVersion":"apps/v1"}}, "resourceVersion":"654", "uid":"084629cf-425f-4d8a-b539-4d8111199ccc", "annotations":map[string]interface {}{"deployment.kubernetes.io/desired-replicas":"1", "deployment.kubernetes.io/max-replicas":"2", "deployment.kubernetes.io/revision":"1"}, "creationTimestamp":"2020-05-06T14:26:04Z", "generation":1, "managedFields":[]interface {}{map[string]interface {}{"fieldsV1":map[string]interface {}{"f:metadata":map[string]interface {}{"f:annotations":map[string]interface {}{"f:deployment.kubernetes.io/revision":map[string]interface {}{}, ".":map[string]interface {}{}, "f:deployment.kubernetes.io/desired-replicas":map[string]interface {}{}, "f:deployment.kubernetes.io/max-replicas":map[string]interface {}{}}, "f:labels":map[string]interface {}{"f:app":map[string]interface {}{}, "f:pod-template-hash":map[string]interface {}{}, ".":map[string]interface {}{}}, "f:ownerReferences":map[string]interface {}{".":map[string]interface {}{}, "k:{\"uid\":\"080eb26a-ec5a-4cbc-b15d-aa63a3d9d252\"}":map[string]interface {}{"f:controller":map[string]interface {}{}, "f:kind":map[string]interface {}{}, "f:name":map[string]interface {}{}, "f:uid":map[string]interface {}{}, ".":map[string]interface {}{}, "f:apiVersion":map[string]interface {}{}, "f:blockOwnerDeletion":map[string]interface {}{}}}}, "f:spec":map[string]interface {}{"f:selector":map[string]interface {}{"f:matchLabels":map[string]interface {}{".":map[string]interface {}{}, "f:app":map[string]interface {}{}, "f:pod-template-hash":map[string]interface {}{}}}, "f:template":map[string]interface {}{"f:metadata":map[string]interface {}{"f:labels":map[string]interface {}{".":map[string]interface {}{}, "f:app":map[string]interface {}{}, "f:pod-template-hash":map[string]interface {}{}}}, "f:spec":map[string]interface {}{"f:containers":map[string]interface {}{"k:{\"name\":\"busybox\"}":map[string]interface {}{"f:name":map[string]interface {}{}, "f:resources":map[string]interface {}{}, "f:terminationMessagePath":map[string]interface {}{}, "f:terminationMessagePolicy":map[string]interface {}{}, ".":map[string]interface {}{}, "f:image":map[string]interface {}{}, "f:imagePullPolicy":map[string]interface {}{}}}, "f:dnsPolicy":map[string]interface {}{}, "f:restartPolicy":map[string]interface {}{}, "f:schedulerName":map[string]interface {}{}, "f:securityContext":map[string]interface {}{}, "f:terminationGracePeriodSeconds":map[string]interface {}{}}}, "f:replicas":map[string]interface {}{}}, "f:status":map[string]interface {}{"f:fullyLabeledReplicas":map[string]interface {}{}, "f:observedGeneration":map[string]interface {}{}, "f:replicas":map[string]interface {}{}}}, "manager":"kube-controller-manager", "operation":"Update", "time":"2020-05-06T14:26:04Z", "apiVersion":"apps/v1", "fieldsType":"FieldsV1"}}, "name":"my-dep-6f9f87db6f", "labels":map[string]interface {}{"app":"my-dep", "pod-template-hash":"6f9f87db6f"}, "namespace":"default", "selfLink":"/apis/apps/v1/namespaces/default/replicasets/my-dep-6f9f87db6f"}, "spec":map[string]interface {}{"selector":map[string]interface {}{"matchLabels":map[string]interface {}{"app":"my-dep", "pod-template-hash":"6f9f87db6f"}}, "template":map[string]interface {}{"metadata":map[string]interface {}{"creationTimestamp":interface {}(nil), "labels":map[string]interface {}{"app":"my-dep", "pod-template-hash":"6f9f87db6f"}}, "spec":map[string]interface {}{"schedulerName":"default-scheduler", "securityContext":map[string]interface {}{}, "terminationGracePeriodSeconds":30, "containers":[]interface {}{map[string]interface {}{"image":"busybox", "imagePullPolicy":"Always", "name":"busybox", "resources":map[string]interface {}{}, "terminationMessagePath":"/dev/termination-log", "terminationMessagePolicy":"File"}}, "dnsPolicy":"ClusterFirst", "restartPolicy":"Always"}}, "replicas":1}, "status":map[string]interface {}{"fullyLabeledReplicas":1, "observedGeneration":1, "replicas":1}, "apiVersion":"apps/v1", "kind":"ReplicaSet"}, map[string]interface {}{"apiVersion":"apps/v1", "kind":"ReplicaSet", "metadata":map[string]interface {}{"generation":1, "managedFields":[]interface {}{map[string]interface {}{"time":"2020-05-06T15:00:09Z", "apiVersion":"apps/v1", "fieldsType":"FieldsV1", "fieldsV1":map[string]interface {}{"f:status":map[string]interface {}{"f:fullyLabeledReplicas":map[string]interface {}{}, "f:observedGeneration":map[string]interface {}{}, "f:replicas":map[string]interface {}{}}}, "manager":"kube-controller-manager", "operation":"Update"}, map[string]interface {}{"apiVersion":"apps/v1", "fieldsType":"FieldsV1", "fieldsV1":map[string]interface {}{"f:spec":map[string]interface {}{"f:replicas":map[string]interface {}{}, "f:selector":map[string]interface {}{"f:matchLabels":map[string]interface {}{".":map[string]interface {}{}, "f:fq-app":map[string]interface {}{}}}, "f:template":map[string]interface {}{"f:metadata":map[string]interface {}{"f:labels":map[string]interface {}{"f:fq-app":map[string]interface {}{}, ".":map[string]interface {}{}}, "f:name":map[string]interface {}{}, "f:namespace":map[string]interface {}{}}, "f:spec":map[string]interface {}{"f:restartPolicy":map[string]interface {}{}, "f:schedulerName":map[string]interface {}{}, "f:securityContext":map[string]interface {}{}, "f:terminationGracePeriodSeconds":map[string]interface {}{}, "f:containers":map[string]interface {}{"k:{\"name\":\"busybox\"}":map[string]interface {}{".":map[string]interface {}{}, "f:image":map[string]interface {}{}, "f:imagePullPolicy":map[string]interface {}{}, "f:name":map[string]interface {}{}, "f:resources":map[string]interface {}{}, "f:terminationMessagePath":map[string]interface {}{}, "f:terminationMessagePolicy":map[string]interface {}{}}}, "f:dnsPolicy":map[string]interface {}{}}}}}, "manager":"kubectl", "operation":"Update", "time":"2020-05-06T15:00:09Z"}}, "name":"test", "namespace":"default", "resourceVersion":"876", "selfLink":"/apis/apps/v1/namespaces/default/replicasets/test", "uid":"90002076-f162-4e88-a2ec-3172d6cb4cb5", "creationTimestamp":"2020-05-06T15:00:09Z"}, "spec":map[string]interface {}{"replicas":1, "selector":map[string]interface {}{"matchLabels":map[string]interface {}{"fq-app":"bb-m01"}}, "template":map[string]interface {}{"metadata":map[string]interface {}{"creationTimestamp":interface {}(nil), "labels":map[string]interface {}{"fq-app":"bb-m01"}, "name":"bb-m01", "namespace":"default"}, "spec":map[string]interface {}{"containers":[]interface {}{map[string]interface {}{"image":"busybox", "imagePullPolicy":"Always", "name":"busybox", "resources":map[string]interface {}{}, "terminationMessagePath":"/dev/termination-log", "terminationMessagePolicy":"File"}}, "dnsPolicy":"ClusterFirst", "restartPolicy":"Always", "schedulerName":"default-scheduler", "securityContext":map[string]interface {}{}, "terminationGracePeriodSeconds":30}}}, "status":map[string]interface {}{"fullyLabeledReplicas":1, "observedGeneration":1, "replicas":1}}}, "kind":"List", "metadata":map[string]interface {}{}, "resourceVersion":"", "selfLink":"", "apiVersion":"v1"}

error: error executing jsonpath "{.items[?(@.spec.replicas==1)].metadata.name}": incompatible types for comparison
```

```
$ kubectl version --client
Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.14-beta.0.11+4fd45574758f6f-dirty", GitCommit:"4fd45574758f6f5b430a77015ed9f1cc16ed1412", GitTreeState:"dirty", BuildDate:"2020-05-06T15:05:44Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"darwin/amd64"}

$ kubectl get rs -ojsonpath='{.items[?(@.spec.replicas==1)].metadata.name}'
my-dep-6f9f87db6f test
```


```
$ kubectl version --client
Client Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.0-alpha.2.294+cab89c7d08a5a9", GitCommit:"cab89c7d08a5a9ae4091a1aaaa9b9c0ab0f307da", GitTreeState:"clean", BuildDate:"2020-05-06T15:08:57Z", GoVersion:"go1.13.4", Compiler:"gc", Platform:"darwin/amd64"}

$ kubectl get rs -ojsonpath='{.items[?(@.spec.replicas==1)].metadata.name}'
my-dep-6f9f87db6f test
```

Several improvements to json-decoding were made to ensure an integer value was decoded into an integer value in memory.

/close
